File: web/src/lib/timezone.ts
// Frontend timezone utilities for ServerSmith.
// All timestamps from API are in UTC ISO8601 format.
// Display must always be in Asia/Shanghai (Beijing time).

const BEIJING_TZ = "Asia/Shanghai";

/**
 * Format a UTC ISO8601 string to Beijing time display string.
 * Example: "2026-05-03T14:00:00Z" → "2026-05-03 22:00:00"
 */
export function formatUTCToBeijing(utcStr: string | null | undefined): string {
  if (!utcStr) return "-";
  try {
    const date = new Date(utcStr);
    return date.toLocaleString("zh-CN", {
      timeZone: BEIJING_TZ,
      year: "numeric",
      month: "2-digit",
      day: "2-digit",
      hour: "2-digit",
      minute: "2-digit",
      second: "2-digit",
      hour12: false,
    });
  } catch {
    return utcStr;
  }
}

/**
 * Format UTC to Beijing date only (no time).
 * Example: "2026-05-02T15:59:59Z" → "2026-05-03"
 */
export function formatUTCToBeijingDate(utcStr: string | null | undefined): string {
  if (!utcStr) return "-";
  try {
    const date = new Date(utcStr);
    return date.toLocaleDateString("zh-CN", {
      timeZone: BEIJING_TZ,
      year: "numeric",
      month: "2-digit",
      day: "2-digit",
    });
  } catch {
    return utcStr;
  }
}

/**
 * Format UTC to relative time (e.g. "2分钟前", "3小时前").
 */
export function timeAgo(utcStr: string | null | undefined): string {
  if (!utcStr) return "-";
  try {
    const now = Date.now();
    const then = new Date(utcStr).getTime();
    const diffSec = Math.floor((now - then) / 1000);

    if (diffSec < 0) return "刚刚";
    if (diffSec < 60) return `${diffSec}秒前`;
    if (diffSec < 3600) return `${Math.floor(diffSec / 60)}分钟前`;
    if (diffSec < 86400) return `${Math.floor(diffSec / 3600)}小时前`;
    return formatUTCToBeijingDate(utcStr);
  } catch {
    return utcStr;
  }
}

File: server/db/timezone.go
package db

import "time"

var TZ *time.Location

func init() {
	var err error
	TZ, err = time.LoadLocation("Asia/Shanghai")
	if err != nil {
		// fallback to UTC+8 fixed offset
		TZ = time.FixedZone("CST", 8*3600)
	}
}

// NowUTC returns current UTC time as RFC3339 string (for DB storage)
func NowUTC() string {
	return time.Now().UTC().Format(time.RFC3339)
}

// NowBeijing returns current Beijing time
func NowBeijing() time.Time {
	return time.Now().In(TZ)
}

// BeijingToUTC converts a Beijing-local time to UTC RFC3339 string
func BeijingToUTC(t time.Time) string {
	// Force interpret as Beijing if not already
	bj := time.Date(t.Year(), t.Month(), t.Day(), t.Hour(), t.Minute(), t.Second(), t.Nanosecond(), TZ)
	return bj.UTC().Format(time.RFC3339)
}

// FormatUTCToBeijing formats a UTC time string to Beijing display string
func FormatUTCToBeijing(utcStr string) string {
	t, err := time.Parse(time.RFC3339, utcStr)
	if err != nil {
		return utcStr
	}
	return t.In(TZ).Format("2006-01-02 15:04:05")
}

// FormatUTCToBeijingShort formats UTC to Beijing date only
func FormatUTCToBeijingShort(utcStr string) string {
	t, err := time.Parse(time.RFC3339, utcStr)
	if err != nil {
		return utcStr
	}
	return t.In(TZ).Format("2006-01-02")
}
